inst/include/dust/dust.hpp

```diff
@@ -363,39 +461,101 @@ class Dust {
     _rng.long_jump();
   }
 
-  void set_data(std::unordered_map<size_t, std::vector<data_t>> data) {
+  void set_data(std::map<size_t, std::vector<data_t>> data) {
     _data = data;
   }
 
   std::vector<real_t> compare_data() {
     std::vector<real_t> res;
     auto d = _data.find(step());
-    // If we don't find data, we will return a vector of length 0
-    // (which we catch and convert to NULL on return to R).
     if (d != _data.end()) {
-      const size_t np = _n_pars == 0 ?
-        _particles.size() : _particles.size() / _n_pars;
       res.resize(_particles.size());
+      compare_data(res, d->second);
+    }
+    return res;
+  }
+
+  void compare_data(std::vector<real_t>& res, const std::vector<data_t>& data) {
+    const size_t np = _particles.size() / n_pars_effective();
 #ifdef _OPENMP
-      #pragma omp parallel for schedule(static) num_threads(_n_threads)
+    #pragma omp parallel for schedule(static) num_threads(_n_threads)
 #endif
-      for (size_t i = 0; i < _particles.size(); ++i) {
-        res[i] = _particles[i].compare_data(d->second[i / np], _rng.state(i));
+    for (size_t i = 0; i < _particles.size(); ++i) {
+      res[i] = _particles[i].compare_data(data[i / np], _rng.state(i));
+    }
+  }
+
+  std::vector<real_t> filter(bool save_history) {
+    if (_data.size() == 0) {
+      throw std::invalid_argument("Data has not been set for this object");
+    }
+
+    const size_t n_particles = _particles.size();
+    const size_t n_particles_each = n_particles / n_pars_effective();
+    std::vector<real_t> log_likelihood(n_pars_effective());
+    std::vector<real_t> log_likelihood_step(n_pars_effective());
+    std::vector<real_t> weights(n_particles);
+    std::vector<size_t> kappa(n_particles);
+
+    if (save_history) {
+      filter_state_.resize(_index.size(), _particles.size(), _data.size());
+      state(filter_state_.history_value_iterator());
+      filter_state_.advance();
+    }
+
+    for (auto & d : _data) {
+      run(d.first);
+      compare_data(weights, d.second);
+
+      // TODO: we should cope better with the case where all weights
+      // are 0; I think that is the behaviour in the model (or rather
+      // the case where there is no data and so we do not resample)
+      //
+      // TODO: we should cope better with the case where one filter
+      // has become impossible but others continue, but that's hard!
+      auto wi = weights.begin();
+      for (size_t i = 0; i < n_pars_effective(); ++i) {
+        log_likelihood_step[i] =
+          scale_log_weights<real_t>(wi, n_particles_each);
+        log_likelihood[i] += log_likelihood_step[i];
+        wi += n_particles_each;
+      }
+
+      // We could move this below if wanted but we'd have to rewrite
+      // the re-sort algorithm.
+      if (save_history) {
+        state(filter_state_.history_value_iterator());
```

Comment:
The history value vs history order meaning and use are a little unclear to me – could you add a few comments to them?